yarn-project/circuit-types/src/l2_block_downloader/l2_block_downloader.ts

```diff
@@ -68,15 +68,35 @@ export class L2BlockDownloader {
   /**
    * Repeatedly queries the block source and adds the received blocks to the block queue.
    * Stops when no further blocks are received.
+   * @param targetBlockNumber - Optional block number to stop at. If given, will download even unproven blocks to get to it.
    * @returns The total number of blocks added to the block queue.
    */
-  private async collectBlocks() {
+  private async collectBlocks(targetBlockNumber?: number) {
     let totalBlocks = 0;
     while (true) {
-      const blocks = await this.l2BlockSource.getBlocks(this.from, 10, this.proven);
+      // If we have a target and have reached it, return
+      if (targetBlockNumber !== undefined && this.from > targetBlockNumber) {
+        log.verbose(`Reached target block number ${targetBlockNumber}`);
+        return totalBlocks;
+      }
+
+      // If we have a target, then request at most the number of blocks to get to it, bypassing 'proven' restriction
+      const [proven, limit] =
+        targetBlockNumber !== undefined ? [false, Math.min(targetBlockNumber - this.from + 1, 10)] : [this.proven, 10];
```

Comment:
I'm finding this surprising behaviour: I created a proven-blocks-only instance, tell it to sync to block `N` and it gave me back a bunch of unproven blocks. Maybe `proven` should be a parameter to this function instead of a property on the instance? Then each client component could chose to set it or not itself.

Reply:
Alternatively, `this.proven` could be the default for this function `collectBlocks(targetBlockOrProven?: number, proven?: boolean)`